cpp/src/arrow/compute/expression.cc

```diff
@@ -1258,6 +1373,92 @@ struct Inequality {
       return call->function_name == "is_valid" ? literal(true) : literal(false);
     }
 
+    if (call->function_name == "is_in") {
+      // Null-matching behavior is complex and reduces the chances of reduction
+      // of `is_in` calls to a single literal for every possible input, so we
+      // abort the simplification if nulls are possible in the input.
+      if (guarantee.nullable) return expr;
+
+      if (!guarantee.bound.is_scalar()) {
+        return Status::Invalid("Cannot simplify inequality on a non-scalar bound");
+      }
+
+      const auto& lhs = Comparison::StripOrderPreservingCasts(call->arguments[0]);
+      if (!lhs.field_ref()) return expr;
+      if (*lhs.field_ref() != guarantee.target) return expr;
+
+      auto options = checked_pointer_cast<SetLookupOptions>(call->options);
+      auto unsimplified_value_set = options->value_set.make_array();
+      if (unsimplified_value_set->length() == 0) return literal(false);
+
+      Type::type value_set_type = unsimplified_value_set->type_id();
+      // Simplification for `is_in` requires that comparison kernels exist, so
+      // we skip simplification for non-primitive and non-base-binary types.
+      if (!is_integer(value_set_type) && !is_temporal(value_set_type)
+          && !is_base_binary_like(value_set_type)) {
+        return expr;
+      }
```

Comment:
You need to figure out a way of making the `switch` you have below work as both dispatching and validation. Trying to keep these two in sync is too error-prone and inefficient/inelegant.